File: app/src/main/java/com/example/sarada/moviereviews/data/FavoriteContentProvider.java
package com.example.sarada.moviereviews.data;

import android.content.ContentProvider;
import android.content.ContentUris;
import android.content.ContentValues;
import android.content.Context;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.example.sarada.moviereviews.data.FavoriteContract.FavoriteEntry;

public class FavoriteContentProvider extends ContentProvider {

    public static final int FAVORITE = 100;
    public static final int FAVORITE_WITH_ID = 101;
    private static final UriMatcher sUriMatcher = buildUriMatcher();

    public static UriMatcher buildUriMatcher() {
        UriMatcher uriMatcher = new UriMatcher(UriMatcher.NO_MATCH);
        uriMatcher.addURI(FavoriteContract.AUTHORITY, FavoriteContract.PATH_FAVORITE, FAVORITE);
        uriMatcher.addURI(FavoriteContract.AUTHORITY, FavoriteContract.PATH_FAVORITE + "/#", FAVORITE_WITH_ID );
        return uriMatcher;
    }

    private FavoriteDbHelper favoriteDbHelper;

    @Override
    public boolean onCreate() {

        Context context = getContext();
        favoriteDbHelper = new FavoriteDbHelper(context);
        return true;
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, String[] projection, String selection, String[] selectionArgs, String sortOrder) {
        final SQLiteDatabase db = favoriteDbHelper.getReadableDatabase();
        int match = sUriMatcher.match(uri);
        Cursor retCursor;
        switch(match){
            case FAVORITE:
                retCursor = db.query(FavoriteEntry.TABLE_NAME, projection, selection, selectionArgs, null, null, sortOrder);
                break;
            default:
                throw new UnsupportedOperationException("Unknown uri:" + uri);
        }
        retCursor.setNotificationUri(getContext().getContentResolver(), uri);
        return retCursor;
    }

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        return null;
    }

    @Nullable
    @Override
    public Uri insert(@NonNull Uri uri, @Nullable ContentValues contentValues) {
        final SQLiteDatabase db = favoriteDbHelper.getWritableDatabase();
        int match = sUriMatcher.match(uri);
        Uri returnUri = null;
        switch(match){
            case FAVORITE:
                long id = db.insert(FavoriteEntry.TABLE_NAME, null, contentValues);
                if(id>0){
                    returnUri = ContentUris.withAppendedId(FavoriteContract.FavoriteEntry.CONTENT_URI, id);
                }else{
                    throw new android.database.SQLException("Failed to insert row into " + uri);
                }
                break;
            case FAVORITE_WITH_ID:
                break;
            default:
                throw new UnsupportedOperationException("Unknown uri:" + uri);
        }
        getContext().getContentResolver().notifyChange(uri, null);
        return returnUri;
    }

    @Override
    public int delete(@NonNull Uri uri, @Nullable String s, @Nullable String[] strings) {
        final SQLiteDatabase db = favoriteDbHelper.getWritableDatabase();
        int match = sUriMatcher.match(uri);
        int favsDeleted;
        switch (match) {
            case FAVORITE_WITH_ID:
                String id = uri.getPathSegments().get(1);
                favsDeleted = db.delete(FavoriteEntry.TABLE_NAME, "movieid=?", new String[]{id});
                break;
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }
        if (favsDeleted != 0) {
            getContext().getContentResolver().notifyChange(uri, null);
        }
        return favsDeleted;
    }

    @Override
    public int update(@NonNull Uri uri, @Nullable ContentValues contentValues, @Nullable String s, @Nullable String[] strings) {
        return 0;
    }
}

File: app/src/main/java/com/example/sarada/moviereviews/database/FavoriteContentProvider.kt
package com.example.sarada.moviereviews.database

import android.content.ContentProvider
import com.example.sarada.moviereviews.database.FavoriteContract.FavoriteEntry
import android.content.ContentValues
import android.content.ContentUris
import android.content.UriMatcher
import android.database.Cursor
import android.database.SQLException
import android.net.Uri
import java.lang.UnsupportedOperationException

class FavoriteContentProvider : ContentProvider() {

    private var favoriteDbHelper: FavoriteDbHelper? = null

    override fun onCreate(): Boolean {
        val context = context
        favoriteDbHelper = FavoriteDbHelper(context)
        return true
    }

    override fun query(
        uri: Uri,
        projection: Array<String>?,
        selection: String?,
        selectionArgs: Array<String>?,
        sortOrder: String?
    ): Cursor {
        val db = favoriteDbHelper!!.readableDatabase
        val retCursor: Cursor = when (sUriMatcher.match(uri)) {
            FAVORITE -> db.query(
                FavoriteEntry.TABLE_NAME,
                projection,
                selection,
                selectionArgs,
                null,
                null,
                sortOrder
            )
            else -> throw UnsupportedOperationException("Unknown uri:$uri")
        }
        retCursor.setNotificationUri(context!!.contentResolver, uri)
        return retCursor
    }

    override fun getType(uri: Uri): String? {
        return null
    }

    override fun insert(uri: Uri, contentValues: ContentValues?): Uri? {
        val db = favoriteDbHelper!!.writableDatabase
        val match = sUriMatcher.match(uri)
        var returnUri: Uri? = null
        when (match) {
            FAVORITE -> {
                val id = db.insert(FavoriteEntry.TABLE_NAME, null, contentValues)
                returnUri = if (id > 0) {
                    ContentUris.withAppendedId(FavoriteEntry.CONTENT_URI, id)
                } else {
                    throw SQLException("Failed to insert row into $uri")
                }
            }
            FAVORITE_WITH_ID -> {}
            else -> throw UnsupportedOperationException("Unknown uri:$uri")
        }
        context!!.contentResolver.notifyChange(uri, null)
        return returnUri
    }

    override fun delete(uri: Uri, s: String?, strings: Array<String>?): Int {
        val db = favoriteDbHelper!!.writableDatabase
        val favsDeleted: Int = when (sUriMatcher.match(uri)) {
            FAVORITE_WITH_ID -> {
                val id = uri.pathSegments[1]
                db.delete(FavoriteEntry.TABLE_NAME, "movieid=?", arrayOf(id))
            }
            else -> throw UnsupportedOperationException("Unknown uri: $uri")
        }
        if (favsDeleted != 0) {
            context!!.contentResolver.notifyChange(uri, null)
        }
        return favsDeleted
    }

    override fun update(
        uri: Uri,
        contentValues: ContentValues?,
        s: String?,
        strings: Array<String>?
    ): Int {
        return 0
    }

    companion object {
        const val FAVORITE = 100
        const val FAVORITE_WITH_ID = 101
        private val sUriMatcher = buildUriMatcher()
        private fun buildUriMatcher(): UriMatcher {
            val uriMatcher = UriMatcher(UriMatcher.NO_MATCH)
            uriMatcher.addURI(FavoriteContract.AUTHORITY, FavoriteContract.PATH_FAVORITE, FAVORITE)
            uriMatcher.addURI(
                FavoriteContract.AUTHORITY,
                FavoriteContract.PATH_FAVORITE + "/#",
                FAVORITE_WITH_ID
            )
            return uriMatcher
        }
    }
}